File: examples/javascript/test-runner.js
// Test Runner for JavaScript Examples
// This script runs all JavaScript examples as integration tests
const fetch = require('node-fetch');
const { spawn } = require('child_process');
const path = require('path');
require('dotenv').config({ path: path.join(__dirname, '..', '.env') });

const BASE_URL = process.env.API_BASE_URL || 'http://localhost:8080';
const API_KEY = process.env.API_BASE_KEY || 'a-test-api-key-from-ekodb';

// Colors for output
const colors = {
    green: '\x1b[32m',
    red: '\x1b[31m',
    yellow: '\x1b[33m',
    blue: '\x1b[34m',
    reset: '\x1b[0m'
};

function log(message, color = 'reset') {
    console.log(`${colors[color]}${message}${colors.reset}`);
}

// Check if server is running
async function checkServer() {
    log('\n=== Checking Server Connection ===', 'blue');
    
    for (let i = 0; i < 30; i++) {
        try {
            const response = await fetch(`${BASE_URL}/api/health`, { timeout: 1000 });
            if (response.ok) {
                log('✓ Server is ready', 'green');
                return true;
            }
        } catch (error) {
            process.stdout.write('.');
            await new Promise(resolve => setTimeout(resolve, 1000));
        }
    }
    
    log('\n✗ Server not responding', 'red');
    return false;
}

// Get authentication token
async function getAuthToken() {
    log('\n=== Getting Authentication Token ===', 'blue');
    
    try {
        const response = await fetch(`${BASE_URL}/api/auth/token`, {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify({ api_key: API_KEY })
        });
        
        if (!response.ok) {
            throw new Error(`Auth failed: ${response.status}`);
        }
        
        const data = await response.json();
        log('✓ Authentication successful', 'green');
        return data.token;
    } catch (error) {
        log(`✗ Authentication failed: ${error.message}`, 'red');
        throw error;
    }
}

// Run a single example file
function runExample(file, token) {
    return new Promise((resolve, reject) => {
        log(`\n=== Running ${file} ===`, 'blue');
        
        const env = {
            ...process.env,
            API_BASE_URL: BASE_URL,
            WS_BASE_URL: process.env.WS_BASE_URL || 'ws://localhost:8080',
            AUTH_TOKEN: token
        };
        
        const child = spawn('node', [file], {
            cwd: path.dirname(file),
            env,
            stdio: 'inherit'
        });
        
        child.on('exit', (code) => {
            if (code === 0) {
                log(`✓ ${path.basename(file)} completed successfully`, 'green');
                resolve();
            } else {
                log(`✗ ${path.basename(file)} failed with code ${code}`, 'red');
                reject(new Error(`Example failed: ${file}`));
            }
        });
        
        child.on('error', (error) => {
            log(`✗ ${path.basename(file)} error: ${error.message}`, 'red');
            reject(error);
        });
    });
}

// Main test runner
async function main() {
    log('\n╔════════════════════════════════════════╗', 'blue');
    log('║  ekoDB JavaScript Examples Test Suite ║', 'blue');
    log('╚════════════════════════════════════════╝', 'blue');
    
    try {
        // Check server
        const serverReady = await checkServer();
        if (!serverReady) {
            process.exit(1);
        }
        
        // Get auth token
        const token = await getAuthToken();
        
        // List of example files to run
        const examples = [
            path.join(__dirname, 'ttl-caching', 'simple_crud.js'),
            path.join(__dirname, 'ttl-caching', 'simple_websocket.js'),
            path.join(__dirname, 'ttl-caching', 'batch_operations.js'),
            path.join(__dirname, 'ttl-caching', 'kv_operations.js'),
            path.join(__dirname, 'ttl-caching', 'collection_management.js'),
            path.join(__dirname, 'ttl-caching', 'document_ttl.js'),
            path.join(__dirname, 'ttl-caching', 'websocket_ttl.js'),
            path.join(__dirname, 'http_functions.js'),
            path.join(__dirname, 'transactions.js')
        ];
        
        log(`\n=== Running ${examples.length} Examples ===`, 'yellow');
        
        let passed = 0;
        let failed = 0;
        
        for (const example of examples) {
            try {
                await runExample(example, token);
                passed++;
            } catch (error) {
                failed++;
                log(`Continuing with next example...`, 'yellow');
            }
            
            // Wait between examples
            await new Promise(resolve => setTimeout(resolve, 2000));
        }
        
        // Summary
        log('\n╔════════════════════════════════════════╗', 'blue');
        log('║           Test Summary                 ║', 'blue');
        log('╚════════════════════════════════════════╝', 'blue');
        log(`Total: ${examples.length}`, 'blue');
        log(`Passed: ${passed}`, 'green');
        log(`Failed: ${failed}`, failed > 0 ? 'red' : 'green');
        
        if (failed > 0) {
            process.exit(1);
        }
        
    } catch (error) {
        log(`\n✗ Test suite failed: ${error.message}`, 'red');
        process.exit(1);
    }
}

// Run if called directly
if (require.main === module) {
    main();
}

module.exports = { checkServer, getAuthToken, runExample };

File: examples/python/test_runner.py
#!/usr/bin/env python3
"""
Test Runner for Python Examples
This script runs all Python examples as integration tests
"""

import asyncio
import os
import sys
import subprocess
import time
from pathlib import Path
import aiohttp
from dotenv import load_dotenv

# Load environment variables
env_path = Path(__file__).parent.parent / ".env"
load_dotenv(env_path)

BASE_URL = os.getenv("API_BASE_URL", "http://localhost:8080")
API_KEY = os.getenv("API_BASE_KEY", "a-test-api-key-from-ekodb")


# Colors for output
class Colors:
    GREEN = "\033[32m"
    RED = "\033[31m"
    YELLOW = "\033[33m"
    BLUE = "\033[34m"
    RESET = "\033[0m"


def log(message, color="RESET"):
    color_code = getattr(Colors, color, Colors.RESET)
    print(f"{color_code}{message}{Colors.RESET}")


async def check_server():
    """Check if server is running"""
    log("\n=== Checking Server Connection ===", "BLUE")

    async with aiohttp.ClientSession() as session:
        for i in range(30):
            try:
                async with session.get(
                    f"{BASE_URL}/api/health", timeout=aiohttp.ClientTimeout(total=1)
                ) as response:
                    if response.status == 200:
                        log("✓ Server is ready", "GREEN")
                        return True
            except Exception:
                sys.stdout.write(".")
                sys.stdout.flush()
                await asyncio.sleep(1)

    log("\n✗ Server not responding", "RED")
    return False


async def get_auth_token():
    """Get authentication token"""
    log("\n=== Getting Authentication Token ===", "BLUE")

    async with aiohttp.ClientSession() as session:
        try:
            async with session.post(
                f"{BASE_URL}/api/auth/token",
                json={"api_key": API_KEY},
                headers={"Content-Type": "application/json"},
            ) as response:
                if response.status != 200:
                    raise Exception(f"Auth failed: {response.status}")

                data = await response.json()
                log("✓ Authentication successful", "GREEN")
                return data["token"]
        except Exception as e:
            log(f"✗ Authentication failed: {e}", "RED")
            raise


def run_example(file_path, token):
    """Run a single example file"""
    log(f"\n=== Running {Path(file_path).name} ===", "BLUE")

    env = os.environ.copy()
    env.update(
        {
            "API_BASE_URL": BASE_URL,
            "WS_BASE_URL": os.getenv("WS_BASE_URL", "ws://localhost:8080"),
            "AUTH_TOKEN": token,
        }
    )

    try:
        result = subprocess.run(
            ["python3", file_path], env=env, capture_output=False, text=True, timeout=60
        )

        if result.returncode == 0:
            log(f"✓ {Path(file_path).name} completed successfully", "GREEN")
            return True
        else:
            log(f"✗ {Path(file_path).name} failed with code {result.returncode}", "RED")
            return False
    except subprocess.TimeoutExpired:
        log(f"✗ {Path(file_path).name} timed out", "RED")
        return False
    except Exception as e:
        log(f"✗ {Path(file_path).name} error: {e}", "RED")
        return False


async def main():
    """Main test runner"""
    log("\n╔════════════════════════════════════════╗", "BLUE")
    log("║   ekoDB Python Examples Test Suite    ║", "BLUE")
    log("╚════════════════════════════════════════╝", "BLUE")

    try:
        # Check server
        server_ready = await check_server()
        if not server_ready:
            sys.exit(1)

        # Get auth token
        token = await get_auth_token()

        # List of example files to run
        # Note: Direct API examples are in the main directory
        # TTL verification tests are in the ttl/ subdirectory
        base_dir = Path(__file__).parent
        examples = [
            base_dir / "simple_crud.py",
            base_dir / "simple_websocket.py",
            base_dir / "http_functions.py",
            base_dir / "batch_operations.py",
            base_dir / "kv_operations.py",
            base_dir / "collection_management.py",
            base_dir / "transactions.py",
            base_dir / "crud_scripts.py",
            base_dir / "ttl" / "document_ttl.py",
            base_dir / "ttl" / "websocket_ttl.py",
        ]

        log(f"\n=== Running {len(examples)} Examples ===", "YELLOW")

        passed = 0
        failed = 0

        for example in examples:
            if example.exists():
                if run_example(str(example), token):
                    passed += 1
                else:
                    failed += 1
                    log("Continuing with next example...", "YELLOW")

                # Wait between examples
                time.sleep(2)
            else:
                log(f"✗ Example not found: {example}", "RED")
                failed += 1

        # Summary
        log("\n╔════════════════════════════════════════╗", "BLUE")
        log("║           Test Summary                 ║", "BLUE")
        log("╚════════════════════════════════════════╝", "BLUE")
        log(f"Total: {len(examples)}", "BLUE")
        log(f"Passed: {passed}", "GREEN")
        log(f"Failed: {failed}", "RED" if failed > 0 else "GREEN")

        if failed > 0:
            sys.exit(1)

    except Exception as e:
        log(f"\n✗ Test suite failed: {e}", "RED")
        sys.exit(1)


if __name__ == "__main__":
    asyncio.run(main())
